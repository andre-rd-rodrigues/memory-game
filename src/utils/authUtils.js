//Storage
const sessionStorage_token = "memory_auth_token";
const sessionStorage_user = "memory_user";

const setToken = (token_arg) =>
  sessionStorage.setItem(sessionStorage_token, token_arg);

let token = () => sessionStorage.getItem(sessionStorage_token);

const setUser = (user) =>
  sessionStorage.setItem(sessionStorage_user, JSON.stringify(user));

const setStorage = (token_arg, user) => {
  setToken(token_arg);
  setUser(user);
};

const getUserFromStorage = JSON.parse(
  sessionStorage.getItem(sessionStorage_user)
);

//Logout

const logout = () => {
  sessionStorage.removeItem(sessionStorage_token);
  window.location.pathname = "/login";
};
export { setStorage, token, setToken, getUserFromStorage, logout };
