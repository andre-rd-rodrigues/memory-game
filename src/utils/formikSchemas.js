import * as yup from "yup";

const dataFields = {
  email: yup
    .string()
    .email("Please enter a valid email")
    .required("Email is required"),
  password: yup
    .string()
    .required("Password is required")
    .min(6, "Password is too short - should be 6 chars minimum.")
    .matches(/[a-zA-Z]/, "Password can only contain letters.")
};

const formikSettings = {
  schemas: {
    loginSchema: yup.object().shape(dataFields),
    registerSchema: yup.object().shape({
      ...dataFields,
      passwordConfirmation: yup
        .string()
        .oneOf([yup.ref("password"), null], "Passwords must match"),
      name: yup
        .string()
        .min(4, "Must be upper than 4 characters")
        .max(100, "Must be lesser than 100 characters")
        .required("Name is required")
    })
  },

  initialValues: {
    login: {
      email: "",
      password: ""
    },
    register: {
      email: "",
      password: "",
      name: ""
    }
  }
};

export default formikSettings;
