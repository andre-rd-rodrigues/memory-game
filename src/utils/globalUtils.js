import { icons } from "mocks/local-data";

const generateNumbersArray = (n) => {
  let result = [];
  for (let i = 1; i <= n; i++) {
    result.push(i);
  }
  return result;
};

const generateRandomIconsBoard = (n) => {
  let result = [];

  for (let i = 0; i < n; i++) {
    let randomIcon = icons[Math.floor(Math.random() * icons.length)];
    let iconsAlreadyPushed = result.filter(
      (icon) => icon === randomIcon
    ).length;
    let isDoubled = iconsAlreadyPushed === 2;

    do {
      randomIcon = icons[Math.floor(Math.random() * icons.length)];

      iconsAlreadyPushed = result.filter((icon) => icon === randomIcon).length;
      isDoubled = iconsAlreadyPushed === 2;
    } while (isDoubled);

    result.push(randomIcon);
  }
  return result;
};
export { generateNumbersArray, generateRandomIconsBoard };
