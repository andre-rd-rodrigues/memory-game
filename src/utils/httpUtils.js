import notification from "./notifications";

//Error message
export const errorMessage = (err) => {
  const message = err?.response?.data;
  notification(message, "error");
  throw message;
};
