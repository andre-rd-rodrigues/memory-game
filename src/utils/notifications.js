import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

const notification = (message, type) =>
  toast[type](<p data-testid="notification">{message}</p>);

export default notification;
