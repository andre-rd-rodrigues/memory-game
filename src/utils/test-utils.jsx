// test-utils.jsx
import React from "react";
import { configureStore } from "@reduxjs/toolkit";
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import reducer from "store/rootReducer";

let store, initialState;

store = configureStore({ reducer, preloadedState: initialState });

initialState = {
  entities: {
    game: {
      settings: {
        theme: "icons",
        boardSize: 4
      },
      board: {
        reset: false
      },
      status: {
        started: false,
        finished: false
      },
      time: {
        started: 0,
        finished: 0
      },
      players: [
        {
          id: 1,
          moves: 0,
          winner: false
        }
      ]
    }
  }
};

const ConnectedComponent = ({ children }) => {
  return (
    <MemoryRouter>
      <Provider store={store}>{children}</Provider>
    </MemoryRouter>
  );
};

function render(ui, { newState, ...renderOptions } = {}) {
  const Wrapper = ({ children }) => {
    if (newState) {
      store = configureStore({ reducer, preloadedState: initialState });
    }
    return (
      <MemoryRouter>
        <Provider store={store}>{children}</Provider>
      </MemoryRouter>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render, initialState, ConnectedComponent };
