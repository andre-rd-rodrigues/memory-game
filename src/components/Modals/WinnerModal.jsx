import React, { useEffect, useState } from "react";
import { postGameMatch } from "api/user";
import Button from "components/Button/Button";
import ListRow from "components/ListRow/ListRow";
import { Modal } from "react-bootstrap";
import { connect } from "react-redux";
import { newGame, restartGame } from "store/entities/game";
import { getTimeElapsed } from "utils/globalUtils";
import "./modals.scss";

const WinnerModal = ({
  show,
  onHide,
  players,
  time,
  restartGame,
  newGame,
  user
}) => {
  const [winner, setWinner] = useState(undefined);
  const [updatedDB, setUpdatedDB] = useState(false);

  const handleGameMatch = async () => {
    //Set winner
    setWinner(players?.filter((player) => player.winner)[0]);

    //Game match
    const match = {
      duration: getTimeElapsed(time.started, time.finished),
      moves: winner?.moves,
      date: new Date().toDateString()
    };

    //Save results in db
    await postGameMatch(match)
      .then(() => setUpdatedDB(true))
      .catch((err) => console.log(err));
  };

  //Finish
  const handleClose = (type) => {
    if (type === "restart") {
      restartGame();
    } else {
      newGame();
    }
    return onHide();
  };

  //Lifecycle
  useEffect(() => {
    if (show && !updatedDB) handleGameMatch();
  }, [show]);

  return (
    <Modal
      show={show}
      onHide={onHide}
      id="winner-modal"
      centered
      backdrop="static"
    >
      <Modal.Body>
        <h2>You won {user.name}!</h2>
        <p id="modal-subtitle">Game over! Here are the results...</p>
        <ListRow
          label="Time Elapsed"
          content={getTimeElapsed(time.started, time.finished)}
        />
        <ListRow label="Moves Taken" content={`${winner?.moves} Moves`} />
        <div className="modal-actions">
          <Button
            label="Restart"
            color="orange"
            onClick={() => handleClose("restart")}
          />
          <Button
            label="New Game"
            color="secondary-light"
            nClick={() => handleClose("newGame")}
          />
        </div>
      </Modal.Body>
    </Modal>
  );
};

const mapStateToProps = (state) => {
  return {
    players: state.entities.game.players,
    time: state.entities.game.time,
    user: state.entities.user
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    restartGame: (obj) => dispatch(restartGame(obj)),
    newGame: (obj) => dispatch(newGame(obj))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(WinnerModal);
