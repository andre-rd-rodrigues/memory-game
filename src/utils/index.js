import * as authUtils from "./authUtils";
import formikSettings from "./formikSchemas";
import * as globalUtils from "./globalUtils";
import * as httpUtils from "./httpUtils";
import notification from "./notifications";

export { authUtils, globalUtils, httpUtils, formikSettings, notification };
