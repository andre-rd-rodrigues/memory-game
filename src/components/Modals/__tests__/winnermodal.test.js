import { mount } from "enzyme";
import { ConnectedComponent } from "utils/test-utils";
import WinnerModal from "../WinnerModal";

describe("List row", () => {
  let props;
  props = {
    label: "Moves Taken",
    content: "10"
  };

  it("should render correctly", () => {
    const wrapper = mount(
      <ConnectedComponent>
        <WinnerModal {...props} />
      </ConnectedComponent>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
