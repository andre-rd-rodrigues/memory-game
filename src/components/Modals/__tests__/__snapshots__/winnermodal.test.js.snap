// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List row should render correctly 1`] = `
<ConnectedComponent>
  <MemoryRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "9qk0e2",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "9qk0e2",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "@@observable": [Function],
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <Connect(WinnerModal)
          content="10"
          label="Moves Taken"
        >
          <WinnerModal
            content="10"
            label="Moves Taken"
            newGame={[Function]}
            players={
              Array [
                Object {
                  "id": 1,
                  "moves": 0,
                  "winner": false,
                },
              ]
            }
            restartGame={[Function]}
            time={
              Object {
                "finished": 0,
                "started": 0,
              }
            }
          >
            <Modal
              animation={true}
              autoFocus={true}
              backdrop="static"
              centered={true}
              dialogAs={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "ModalDialog",
                  "render": [Function],
                }
              }
              enforceFocus={true}
              id="winner-modal"
              keyboard={true}
              restoreFocus={true}
              show={false}
            >
              <Modal
                autoFocus={true}
                backdrop="static"
                backdropTransition={[Function]}
                containerClassName="modal-open"
                enforceFocus={true}
                keyboard={true}
                manager={
                  BootstrapModalManager {
                    "containers": Array [],
                    "data": Array [],
                    "handleContainerOverflow": true,
                    "hideSiblingNodes": true,
                    "modals": Array [],
                    "scrollbarSize": 0,
                  }
                }
                onEnter={[Function]}
                onEntering={[Function]}
                onEscapeKeyDown={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                renderBackdrop={[Function]}
                renderDialog={[Function]}
                restoreFocus={true}
                show={false}
                transition={[Function]}
              />
            </Modal>
          </WinnerModal>
        </Connect(WinnerModal)>
      </Provider>
    </Router>
  </MemoryRouter>
</ConnectedComponent>
`;
